File: src/main/java/_02_functional_but_no_good/Sample.java
package _02_functional_but_no_good;

import java.util.List;

import static util.AirportNameUtil.getNameOfAirport;

public class Sample {
    public static String getNameOfAirport2(String iata) {
        try {
            return getNameOfAirport(iata);
        } catch (Exception ex) {
            throw new RuntimeException(ex);
        }
    }

    public static void main(String[] args) {
        List<String> iataCodes = List.of("AUS", "IAH", "DFW", "TAS", "SAT");

        iataCodes.stream()
            .map(Sample::getNameOfAirport2)
            .forEach(System.out::println);
        //curl up in a corner and cry

        //don't do this
    }
}

File: src/main/kotlin/_09_functional_deal_with_it_kotlin_and_multiple_exception_handling/Sample.kt
package _09_functional_deal_with_it_kotlin_and_multiple_exception_handling

import io.vavr.control.Try
import java.io.IOException

fun main() {
    val iataCodes = listOf("AUS", "IAH", "DFW", "TAS", "SAT")

    // There will be 3 transformation
    // first transform: uppercase but may be exception
        // fallback logic of first: ignore if there is an exception
    // second transform: lowercase but may be exception
        // fallback logic of second: print "error: {error message}" and ignore it if there is an exception
    // third transform: reversed
        // fallback logic of third: x
    // final: print the result

    iataCodes.asSequence()

        // first transform
        .map { Try.of { firstTransformMayThrowException(it) } }
        // first fallback logic
        .filter { it.isSuccess }

        // second transform
        .map { it.map { result1 -> secondTransformMayThrowException(result1) } }
        // second fallback logic
        .map { it.recover { ex -> println("error: ${ex.message}"); null } }.filter { it.get() != null }

        // third transform
        .map { it.map { result2 -> thirdTransformPureFunction(result2) } }

        // final
        .forEach {
            if (it.get().length == 3) {
                println("Final result: ${it.get()}")
            } else {
                throw AssertionError("it should be 3 characters: $it")
            }
        }
}

@Throws(IOException::class)
private fun firstTransformMayThrowException(code: String): String {
    if (Math.random() > 0.5) {
        println("[transformMayThrowException1] code: $code")
        throw IOException("[transformMayThrowException1] IOException")
    }

    return code.uppercase()
}
private fun secondTransformMayThrowException(code: String): String {
    if (Math.random() > 0.5) {
        // println("[transformMayThrowException2] code: $code")
        throw ArithmeticException("[transformMayThrowException2] ArithmeticException: $code")
    }

    return code.lowercase()
}

private fun thirdTransformPureFunction(code: String): String {
    return code.reversed()
}
